### PR TITLE
Ubuntu Platform, part two.

### PR DESCRIPTION
**Pull request by pimvanpelt**
Part two of the Ubuntu platform.
*    Added mgos_uptime_micros()
*    Added ability to Makefile to build lib.a
     *    Also correctly sequence generated files before app sources (so mgos_config.ch exist)
*    Mongoose now runs in a child process, separated from the main process.
     *    Added an IPC socketpair between main and mongoose
     *    When Mongoose chroot()s, it uses IPC to open files in the host filesystem
     *    Mongoose WDT updates the main process
     *    Main process will terminate mongoose on WDT timeout.
*    Added capabilities (cap_sys_chroot and cap_net_bind_service)
*    Drop privileges in Mongoose task after chroot()ing.
    *    The application can run as root, in which case privs are dropped upon start
    *    The application can run as user with capabilities `cap_sys_chroot,cap_net_bind_service+ep`
*    Add required flags (user, group, chroot) for the main application.
*    Add LOGM() with a color coding to distinguish mongoose logging from main logging.
*    Switch to clang-format.

The project is now able to compile `mongoose` and `i2c` lib.a correctly.

```
rm -rf build/
mos build --no-platform-check --local --verbose --platform ubuntu
sudo setcap "cap_sys_chroot,cap_net_bind_service+ep" build/bin/empty
build/bin/empty 

main                 PIDs: parent=15671 child=15672 uid=1000 gid=1000 euid=1000 egid=1000
ubuntu_cap_init      Setting chroot=/home/pim/src/empty/build/fs
ubuntu_cap_init      pid=15672 uid=1000 gid=1000 euid=1000 egid=1000
mongoose_init        Mongoose OS 201901241501 (20190124-150158/2.10.2-43-g038f7765-master)
mongoose_init        empty 1.0 (20190124-150158/2.10.0-dirty-master-g01d3f5b73)
mongoose_init        CPU: 1605 MHz, heap: 16799997952 total, 5432848384 free
mongoose_init        Network: ip=194.126.235.37 netmask=255.255.255.240 gateway=194.126.235.33
mgos_hw_timers_init  Not implemented yet
mgos_vfs_common_init Not implemented yet
mgos_core_fs_init    Not implemented yet
mgos_gpio_hal_init   Not implemented yet
mgos_sys_config_init MAC: 0002C94F0668
mgos_sys_config_init WDT: 30 seconds
mgos_deps_init       init i2c...
mgos_i2c_create      I2C init ok (dev: /dev/i2c-6, freq: 100000)
mgos_init            Init done, RAM: 16799997952 total, 5432848384 free, 0 min free
mgos_event_trigger   ev MOS0 triggered 0 handlers
app_cb               counter=1 uptime=1019778
i2cscan              Found device at I2C address 0x48
```

### CODE REVIEW

**Comment by pimvanpelt** [comment]
Perhaps hold off on review/merging this. There are still some issues with generating artifacts (like json0.conf and a complete config schema). Will ping once I made more progress.

**Comment by pimvanpelt** [comment]
Ready to merge. PTAL @rojer

**Comment by pimvanpelt** [comment]
5eb7cae needs a release of mgos/ubuntu-build:r2 as we're now linking in libcap, and I've made it a static executable for portabliity so users do not need to install libcap-dev Ubuntu package on their host.

File: fw/platforms/ubuntu/src/ubuntu.h [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "mgos.h"
+#include "mgos_system.h"
+
```
Comment (rojer) [inline fw/platforms/ubuntu/src/ubuntu.h]: `extern "C"` in the headers

File: fw/platforms/ubuntu/src/ubuntu.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "mgos.h"
```
Comment (rojer) [inline fw/platforms/ubuntu/src/ubuntu.h]: please don't include mgos.h in header files. it's a big bucket and brings in a lot of stuff that users may not want. IWYU.

File: fw/platforms/ubuntu/src/ubuntu_cap.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
+#include <sys/capability.h>
```
Comment (rojer) [inline fw/platforms/ubuntu/src/ubuntu_cap.c]: copyright blob missing

File: fw/platforms/ubuntu/src/ubuntu.h (hunk 1 above)
Comment (pimvanpelt) [reply fw/platforms/ubuntu/src/ubuntu.h]: Done.

File: fw/platforms/ubuntu/src/ubuntu_cap.c (hunk 3 above)
Comment (pimvanpelt) [reply fw/platforms/ubuntu/src/ubuntu_cap.c]: Added.

**Comment by rojer** [comment]
integrating this now, thanks a lot!

**Comment by rojer** [comment]
https://github.com/cesanta/mongoose-os/commit/324ab2516a41196ac55d6ee15e39ac07374e7e66

**Pull request closed without merging**